<!DOCTYPE html>
<html>
<head>
<title>AW Docs</title>
<style>
	a:link {text-decoration: none;color:#6666CC;}
	a.white {font-family:verdana; font-size:12px; color:white; text-decoration: underline;}
	a.current {font-family:verdana; font-size:12px; color:white; text-decoration: none;}
	a.ext {font-weight: bold;}
	mark.white {font-family:verdana; font-size:12px; color:white;background:none;}
	.tx    {border: 0px; margin-left: auto; margin-right: auto;}
    .tx td {border: 0px; padding: 3px; }
	body {font-family: Verdana, Arial, Helvetica, sans-serif;  font-size: 14px; }
</style>
</head>

<body>
<a id="top"></a>
<table style="width: 800px; border: 2px solid #999999; padding: 0; border-spacing:0; border-collapse:collapse; margin-left: auto; margin-right: auto;">
	<tr><td>
		<table style="width: 100%; border: 0px; padding: 0;  border-collapse:collapse;">
			<tr>
				<td style="text-align: left; vertical-align: top; padding: 0">
					<a href="http://www.agcol.arizona.edu">
						<img src="./img/agcol_logo.gif" alt="AGCoL" style="width: 180px; border:1px solid black"></a>
				</td>
				<td style="padding-bottom:5px; padding-left:40px; text-align: left;  font-family: Verdana, Arial; font-size: 20px; font-weight: bold;color:MidnightBlue">
					AW Documentation
				</td>
				<td style="text-align: right; vertical-align: top; padding: 0">
					<a href="http://www.arizona.edu">
						<img src="./img/UA.gif" alt="UA" title="University of Arizona"  style="width: 40px; border:1px solid black"></a>
					<br>
					<a href="http://www.bio5.org">
						<img src="./img/bio5.gif" alt="BIO5" title="BIO5"  style="width: 40px; border:1px solid black"></a>
				</td>
			</tr>
			<tr>
				<td colspan="3" style="text-align: left; vertical-align: top; padding: 0">
				<table style="width: 100%; padding: 0;  border-spacing:0; border-collapse:collapse;">
					<tr>
						<td style="background-color: #666666; padding-bottom:4px;">
						<a href="https://csoderlund.github.io/AW" class="white">AW Home</a> <mark class="white">|</mark>
						<a href="https://github.com/csoderlund/AW/releases" class="white">Download</a> <mark class="white">|</mark>
						<a href="./Docs.html" class="current">Docs</a> <mark class="white">|</mark>
						<a href="./Tour.html" class="white">Tour</a>
					</tr>
				</table>
				</td>
			</tr>
		</table>
	</td></tr>

	<tr><td style='vertical-align: top; text-align: left; padding:15px;'>
	<!---- START BODY -->

<table class="tx">
	<tr>
		<td><a href=index.html>Main Page</a></td>

		<td>--</td>
	</tr>
	<tr>
		<td> <a href=release.html>Release Notes</a></td>
		<td>--</td>
	</tr>
	<tr>
		<td> <a href=Overview.html>Overview</a></td>
		<td>  Inputs, processing, outputs</td>
	</tr>
	<tr>
		<td> <a href=Tour.html>Tour</a></td>
		<td>  Snapshots of processing and results</td>
	</tr>

	<tr><td colspan="2"></td>
	<tr><td colspan="2"><b>Main documentation</b></td></tr>
	<tr>
		<td> <a href=Pipeline.html>Pipeline</a></td>
		<td> Running the pipeline
	</tr>
	<tr>
		<td> <a href=Files.html>Files</a></td>
		<td> File formats and naming convensions
	</tr>
	<tr>
		<td><a href=runAW.html>runAW</a></td>
		<td>Building the AW database</td>
	</tr>
	<td>&nbsp;&nbsp;&nbsp;</td>
	<tr><td colspan=2>The <tt>runAW</tt> interface has the above three main documentations available from it.</td></tr>
	<tr><td colspan=2>To query and view the database, just type: <tt>viewAW &lt;database name&gt;</tt>.</td></tr>
	<tr><td colspan=2>The <tt>viewAW</tt> interface has "Help" on every display page.</td></tr>

</table>

<h2>References</h2>

<u>Pipeline:</u>
<ol>
<li>FastQC: A quality control tool for high throughput sequence data. [bioinformatics.babraham.ac.uk/projects/fastqc]
<li>Bolger AM, Lohse M, Usadel B (2014) Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics.
<li>Trapnell C, Pachter L, Salzberg SL (2009) TopHat: discovering splice junctions with RNA-Seq. Bioinformatics 25: 1105-1111.

<li>Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R (2009) The Sequence Alignment/Map format and SAMtools. Bioinformatics 25: 2078-2079.

<li>Quinlan AR, Hall IM (2010) BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics 26: 841-842.

<li>Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P, Chaisson M, Gingeras TR (2013) STAR: ultrafast universal RNA-seq aligner. Bioinformatics 29: 15-21.

<li>Roberts A, Pachter L (2013) Streaming fragment assignment for real-time analysis of sequencing experiments. Nat Methods 10: 71-73.

</ol>
<u>runAW</u>: the output of the following software can be input for variant effects.
<ol>

 <li>Cingolani P, Platts A, Wang le L, Coon M, Nguyen T, Wang L, Land SJ, Lu X, Ruden DM (2012) A program for annotating and predicting the effects of single nucleotide polymorphisms, SnpEff: SNPs in the genome of Drosophila melanogaster strain w1118; iso-2; iso-3. Fly (Austin) 6: 80-92.
			<li>McLaren W, Pritchard B, Rios D, Chen Y, Flicek P, Cunningham F (2010) Deriving the consequences of genomic variants with the Ensembl API and SNP Effect Predictor. Bioinformatics 26: 2069-2070.
</ol>

	<!---- END BODY -->
	</td></tr>
	<tr><td style="background-color: #DDDDDD; text-align: center; padding: 0; font-family: Verdana, Arial; font-size: 12px;">
                        Email Comments To: <a href="mailto:tcw@agcol.arizona.edu">tcw@agcol.arizona.edu</a>

</td></tr>
</table>
</body>
</html>

